In [76]:
import pandas as pd
import numpy as np
from datetime import datetime, timedelta
import warnings
warnings.filterwarnings('ignore')

In [77]:
from google.colab import files
uploaded = files.upload()

Saving student.csv to student (2).csv


In [78]:
df = pd.read_csv('student.csv')

df

รหัสนักศึกษา      ชื่อ   อายุ     เพศ             คณะ   GPA  ส่วนสูง_cm  \
0          64001     สมชาย   20.0     ชาย     วิทยาศาสตร์  3.21       170.0   
1          64002    สมหญิง   21.0    หญิง           วิทย์  3.45       165.0   
2          64003       NaN    NaN    Male         Science  3.67       175.0   
3          64004  สมศักดิ์   22.0       ช  คณะวิทยาศาสตร์ -1.00       168.0   
4          64005      สมใจ   19.0       ญ           วิศวะ  3.89       160.0   
5          64006     กานต์  150.0     ชาย     Engineering  3.12       172.0   
6          64007     กิตติ   21.0    หญิง        วิศวกรรม  4.50       158.0   
7          64008      จิรา   20.0       F      ศิลปศาสตร์  2.98       180.0   
8          64009    ชนาธิป   23.0       M            Arts  3.34       163.0   
9          64010      ดารา   21.0    หญิง           ศิลป์  3.56       167.0   
10         64011      ธีระ    NaN     NaN             NaN  3.78         NaN   
11         64012       นภา   22.0     ชาย           วิทย์   NaN       171.0   
12         64013     บุญมี   20.0    หญิง           วิศวฯ  3.23       169.0   
13         64014     ปรีชา   21.0     ชาย    Liberal Arts  3.45       164.0   
14         64015     พิมพ์   22.0  Female             Sci  3.67       162.0   
15         64016      ภูมิ   19.0       ช     วิทยาศาสตร์  3.89       173.0   
16         64017      มานี   20.0       ญ  วิศวกรรมศาสตร์  3.01       166.0   
17         64018      ยุพา   21.0    หญิง             NaN  3.23       161.0   
18         64019     รัตนา    NaN     ชาย      ศิลปศาสตร์   NaN       174.0   
19         64020     ลัดดา   22.0     NaN         Science  3.45       168.0   

    น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์  
0         65.0                     15.0  
1         58.0                     20.0  
2          NaN                     25.0  
3         70.0                     10.0  
4         52.0                     30.0  
5         75.0                     12.0  
6         48.0                     35.0  
7         82.0                     18.0  
8         55.0                     22.0  
9         60.0                     28.0  
10        68.0                      NaN  
11         NaN                     16.0  
12        72.0                     24.0  
13        54.0                     19.0  
14        56.0                     26.0  
15        78.0                     14.0  
16        62.0                     21.0  
17        50.0                     27.0  
18        76.0                      NaN  
19        64.0                     23.0

# **3.1 การจัดการข้อมูลหาย**

วิธีที่ 1: ลบแถวที่มีค่าหาย

In [83]:
df_dropna = df.dropna()
print(f"วิธีที่ 1: ลบแถวที่มีค่าหาย (จาก {len(df)} แถว เหลือ {len(df_dropna)} แถว)")

df_dropna

วิธีที่ 1: ลบแถวที่มีค่าหาย (จาก 20 แถว เหลือ 14 แถว)


รหัสนักศึกษา      ชื่อ   อายุ     เพศ             คณะ   GPA  ส่วนสูง_cm  \
0          64001     สมชาย   20.0     ชาย     วิทยาศาสตร์  3.21       170.0   
1          64002    สมหญิง   21.0    หญิง           วิทย์  3.45       165.0   
3          64004  สมศักดิ์   22.0       ช  คณะวิทยาศาสตร์ -1.00       168.0   
4          64005      สมใจ   19.0       ญ           วิศวะ  3.89       160.0   
5          64006     กานต์  150.0     ชาย     Engineering  3.12       172.0   
6          64007     กิตติ   21.0    หญิง        วิศวกรรม  4.50       158.0   
7          64008      จิรา   20.0       F      ศิลปศาสตร์  2.98       180.0   
8          64009    ชนาธิป   23.0       M            Arts  3.34       163.0   
9          64010      ดารา   21.0    หญิง           ศิลป์  3.56       167.0   
12         64013     บุญมี   20.0    หญิง           วิศวฯ  3.23       169.0   
13         64014     ปรีชา   21.0     ชาย    Liberal Arts  3.45       164.0   
14         64015     พิมพ์   22.0  Female             Sci  3.67       162.0   
15         64016      ภูมิ   19.0       ช     วิทยาศาสตร์  3.89       173.0   
16         64017      มานี   20.0       ญ  วิศวกรรมศาสตร์  3.01       166.0   

    น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์  
0         65.0                     15.0  
1         58.0                     20.0  
3         70.0                     10.0  
4         52.0                     30.0  
5         75.0                     12.0  
6         48.0                     35.0  
7         82.0                     18.0  
8         55.0                     22.0  
9         60.0                     28.0  
12        72.0                     24.0  
13        54.0                     19.0  
14        56.0                     26.0  
15        78.0                     14.0  
16        62.0                     21.0

วิธีที่ 2 เติมค่าด้วยค่าเฉลี่ย

In [102]:
# แสดงข้อมูลก่อนเติมค่า
print("ก่อนเติมค่าเฉลี่ยของอายุ")

df

ก่อนเติมค่าเฉลี่ยของอายุ


รหัสนักศึกษา      ชื่อ        อายุ     เพศ             คณะ   GPA  \
0          64001     สมชาย   20.000000     ชาย     วิทยาศาสตร์  3.21   
1          64002    สมหญิง   21.000000    หญิง           วิทย์  3.45   
2          64003       NaN   28.470588    Male         Science  3.67   
3          64004  สมศักดิ์   22.000000       ช  คณะวิทยาศาสตร์ -1.00   
4          64005      สมใจ   19.000000       ญ           วิศวะ  3.89   
5          64006     กานต์  150.000000     ชาย     Engineering  3.12   
6          64007     กิตติ   21.000000    หญิง        วิศวกรรม  4.50   
7          64008      จิรา   20.000000       F      ศิลปศาสตร์  2.98   
8          64009    ชนาธิป   23.000000       M            Arts  3.34   
9          64010      ดารา   21.000000    หญิง           ศิลป์  3.56   
10         64011      ธีระ   28.470588     NaN         Science  3.78   
11         64012       นภา   22.000000     ชาย           วิทย์   NaN   
12         64013     บุญมี   20.000000    หญิง           วิศวฯ  3.23   
13         64014     ปรีชา   21.000000     ชาย    Liberal Arts  3.45   
14         64015     พิมพ์   22.000000  Female             Sci  3.67   
15         64016      ภูมิ   19.000000       ช     วิทยาศาสตร์  3.89   
16         64017      มานี   20.000000       ญ  วิศวกรรมศาสตร์  3.01   
17         64018      ยุพา   21.000000    หญิง         Science  3.23   
18         64019     รัตนา   28.470588     ชาย      ศิลปศาสตร์   NaN   
19         64020     ลัดดา   22.000000     NaN         Science  3.45   

    ส่วนสูง_cm  น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์     คณะ_มาตรฐาน  
0        170.0        65.0                     15.0     วิทยาศาสตร์  
1        165.0        58.0                     20.0     วิทยาศาสตร์  
2        175.0         NaN                     25.0     วิทยาศาสตร์  
3        168.0        70.0                     10.0     วิทยาศาสตร์  
4        160.0        52.0                     30.0           วิศวะ  
5        172.0        75.0                     12.0     Engineering  
6        158.0        48.0                     35.0        วิศวกรรม  
7        180.0        82.0                     18.0      ศิลปศาสตร์  
8        163.0        55.0                     22.0      ศิลปศาสตร์  
9        167.0        60.0                     28.0      ศิลปศาสตร์  
10         NaN        68.0                      NaN     วิทยาศาสตร์  
11       171.0         NaN                     16.0     วิทยาศาสตร์  
12       169.0        72.0                     24.0           วิศวฯ  
13       164.0        54.0                     19.0    Liberal Arts  
14       162.0        56.0                     26.0             Sci  
15       173.0        78.0                     14.0     วิทยาศาสตร์  
16       166.0        62.0                     21.0  วิศวกรรมศาสตร์  
17       161.0        50.0                     27.0     วิทยาศาสตร์  
18       174.0        76.0                      NaN      ศิลปศาสตร์  
19       168.0        64.0                     23.0     วิทยาศาสตร์

In [103]:
# เหมาะกับข้อมูลตัวเลข - ใช้ค่าเฉลี่ย
mean_age = df['อายุ'].mean()
df['อายุ'].fillna(mean_age, inplace=True)

print(f"\nค่าเฉลี่ยอายุ - {mean_age:.1f} ปี")
print("หลังเติมค่าเฉลี่ยของอายุ")

df


ค่าเฉลี่ยอายุ - 28.5 ปี
หลังเติมค่าเฉลี่ยของอายุ


รหัสนักศึกษา      ชื่อ        อายุ     เพศ             คณะ   GPA  \
0          64001     สมชาย   20.000000     ชาย     วิทยาศาสตร์  3.21   
1          64002    สมหญิง   21.000000    หญิง           วิทย์  3.45   
2          64003       NaN   28.470588    Male         Science  3.67   
3          64004  สมศักดิ์   22.000000       ช  คณะวิทยาศาสตร์ -1.00   
4          64005      สมใจ   19.000000       ญ           วิศวะ  3.89   
5          64006     กานต์  150.000000     ชาย     Engineering  3.12   
6          64007     กิตติ   21.000000    หญิง        วิศวกรรม  4.50   
7          64008      จิรา   20.000000       F      ศิลปศาสตร์  2.98   
8          64009    ชนาธิป   23.000000       M            Arts  3.34   
9          64010      ดารา   21.000000    หญิง           ศิลป์  3.56   
10         64011      ธีระ   28.470588     NaN         Science  3.78   
11         64012       นภา   22.000000     ชาย           วิทย์   NaN   
12         64013     บุญมี   20.000000    หญิง           วิศวฯ  3.23   
13         64014     ปรีชา   21.000000     ชาย    Liberal Arts  3.45   
14         64015     พิมพ์   22.000000  Female             Sci  3.67   
15         64016      ภูมิ   19.000000       ช     วิทยาศาสตร์  3.89   
16         64017      มานี   20.000000       ญ  วิศวกรรมศาสตร์  3.01   
17         64018      ยุพา   21.000000    หญิง         Science  3.23   
18         64019     รัตนา   28.470588     ชาย      ศิลปศาสตร์   NaN   
19         64020     ลัดดา   22.000000     NaN         Science  3.45   

    ส่วนสูง_cm  น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์     คณะ_มาตรฐาน  
0        170.0        65.0                     15.0     วิทยาศาสตร์  
1        165.0        58.0                     20.0     วิทยาศาสตร์  
2        175.0         NaN                     25.0     วิทยาศาสตร์  
3        168.0        70.0                     10.0     วิทยาศาสตร์  
4        160.0        52.0                     30.0           วิศวะ  
5        172.0        75.0                     12.0     Engineering  
6        158.0        48.0                     35.0        วิศวกรรม  
7        180.0        82.0                     18.0      ศิลปศาสตร์  
8        163.0        55.0                     22.0      ศิลปศาสตร์  
9        167.0        60.0                     28.0      ศิลปศาสตร์  
10         NaN        68.0                      NaN     วิทยาศาสตร์  
11       171.0         NaN                     16.0     วิทยาศาสตร์  
12       169.0        72.0                     24.0           วิศวฯ  
13       164.0        54.0                     19.0    Liberal Arts  
14       162.0        56.0                     26.0             Sci  
15       173.0        78.0                     14.0     วิทยาศาสตร์  
16       166.0        62.0                     21.0  วิศวกรรมศาสตร์  
17       161.0        50.0                     27.0     วิทยาศาสตร์  
18       174.0        76.0                      NaN      ศิลปศาสตร์  
19       168.0        64.0                     23.0     วิทยาศาสตร์

วิธีที่ 3 เติมค่าด้วยค่าที่พบบ่อยที่สุด

In [104]:
# แสดงข้อมูลก่อนเติมค่า
print("ก่อนเติมค่าคณะที่พบบ่อยที่สุด")

df

ก่อนเติมค่าคณะที่พบบ่อยที่สุด


รหัสนักศึกษา      ชื่อ        อายุ     เพศ             คณะ   GPA  \
0          64001     สมชาย   20.000000     ชาย     วิทยาศาสตร์  3.21   
1          64002    สมหญิง   21.000000    หญิง           วิทย์  3.45   
2          64003       NaN   28.470588    Male         Science  3.67   
3          64004  สมศักดิ์   22.000000       ช  คณะวิทยาศาสตร์ -1.00   
4          64005      สมใจ   19.000000       ญ           วิศวะ  3.89   
5          64006     กานต์  150.000000     ชาย     Engineering  3.12   
6          64007     กิตติ   21.000000    หญิง        วิศวกรรม  4.50   
7          64008      จิรา   20.000000       F      ศิลปศาสตร์  2.98   
8          64009    ชนาธิป   23.000000       M            Arts  3.34   
9          64010      ดารา   21.000000    หญิง           ศิลป์  3.56   
10         64011      ธีระ   28.470588     NaN         Science  3.78   
11         64012       นภา   22.000000     ชาย           วิทย์   NaN   
12         64013     บุญมี   20.000000    หญิง           วิศวฯ  3.23   
13         64014     ปรีชา   21.000000     ชาย    Liberal Arts  3.45   
14         64015     พิมพ์   22.000000  Female             Sci  3.67   
15         64016      ภูมิ   19.000000       ช     วิทยาศาสตร์  3.89   
16         64017      มานี   20.000000       ญ  วิศวกรรมศาสตร์  3.01   
17         64018      ยุพา   21.000000    หญิง         Science  3.23   
18         64019     รัตนา   28.470588     ชาย      ศิลปศาสตร์   NaN   
19         64020     ลัดดา   22.000000     NaN         Science  3.45   

    ส่วนสูง_cm  น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์     คณะ_มาตรฐาน  
0        170.0        65.0                     15.0     วิทยาศาสตร์  
1        165.0        58.0                     20.0     วิทยาศาสตร์  
2        175.0         NaN                     25.0     วิทยาศาสตร์  
3        168.0        70.0                     10.0     วิทยาศาสตร์  
4        160.0        52.0                     30.0           วิศวะ  
5        172.0        75.0                     12.0     Engineering  
6        158.0        48.0                     35.0        วิศวกรรม  
7        180.0        82.0                     18.0      ศิลปศาสตร์  
8        163.0        55.0                     22.0      ศิลปศาสตร์  
9        167.0        60.0                     28.0      ศิลปศาสตร์  
10         NaN        68.0                      NaN     วิทยาศาสตร์  
11       171.0         NaN                     16.0     วิทยาศาสตร์  
12       169.0        72.0                     24.0           วิศวฯ  
13       164.0        54.0                     19.0    Liberal Arts  
14       162.0        56.0                     26.0             Sci  
15       173.0        78.0                     14.0     วิทยาศาสตร์  
16       166.0        62.0                     21.0  วิศวกรรมศาสตร์  
17       161.0        50.0                     27.0     วิทยาศาสตร์  
18       174.0        76.0                      NaN      ศิลปศาสตร์  
19       168.0        64.0                     23.0     วิทยาศาสตร์

In [105]:
# เหมาะกับข้อมูลประเภท - ใช้ค่าที่พบบ่อยที่สุด (mode)
df['คณะ'].fillna(df['คณะ'].mode()[0], inplace=True)

print("\nหลังเติมค่าคณะที่พบบ่อยที่สุด")

df


หลังเติมค่าคณะที่พบบ่อยที่สุด


รหัสนักศึกษา      ชื่อ        อายุ     เพศ             คณะ   GPA  \
0          64001     สมชาย   20.000000     ชาย     วิทยาศาสตร์  3.21   
1          64002    สมหญิง   21.000000    หญิง           วิทย์  3.45   
2          64003       NaN   28.470588    Male         Science  3.67   
3          64004  สมศักดิ์   22.000000       ช  คณะวิทยาศาสตร์ -1.00   
4          64005      สมใจ   19.000000       ญ           วิศวะ  3.89   
5          64006     กานต์  150.000000     ชาย     Engineering  3.12   
6          64007     กิตติ   21.000000    หญิง        วิศวกรรม  4.50   
7          64008      จิรา   20.000000       F      ศิลปศาสตร์  2.98   
8          64009    ชนาธิป   23.000000       M            Arts  3.34   
9          64010      ดารา   21.000000    หญิง           ศิลป์  3.56   
10         64011      ธีระ   28.470588     NaN         Science  3.78   
11         64012       นภา   22.000000     ชาย           วิทย์   NaN   
12         64013     บุญมี   20.000000    หญิง           วิศวฯ  3.23   
13         64014     ปรีชา   21.000000     ชาย    Liberal Arts  3.45   
14         64015     พิมพ์   22.000000  Female             Sci  3.67   
15         64016      ภูมิ   19.000000       ช     วิทยาศาสตร์  3.89   
16         64017      มานี   20.000000       ญ  วิศวกรรมศาสตร์  3.01   
17         64018      ยุพา   21.000000    หญิง         Science  3.23   
18         64019     รัตนา   28.470588     ชาย      ศิลปศาสตร์   NaN   
19         64020     ลัดดา   22.000000     NaN         Science  3.45   

    ส่วนสูง_cm  น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์     คณะ_มาตรฐาน  
0        170.0        65.0                     15.0     วิทยาศาสตร์  
1        165.0        58.0                     20.0     วิทยาศาสตร์  
2        175.0         NaN                     25.0     วิทยาศาสตร์  
3        168.0        70.0                     10.0     วิทยาศาสตร์  
4        160.0        52.0                     30.0           วิศวะ  
5        172.0        75.0                     12.0     Engineering  
6        158.0        48.0                     35.0        วิศวกรรม  
7        180.0        82.0                     18.0      ศิลปศาสตร์  
8        163.0        55.0                     22.0      ศิลปศาสตร์  
9        167.0        60.0                     28.0      ศิลปศาสตร์  
10         NaN        68.0                      NaN     วิทยาศาสตร์  
11       171.0         NaN                     16.0     วิทยาศาสตร์  
12       169.0        72.0                     24.0           วิศวฯ  
13       164.0        54.0                     19.0    Liberal Arts  
14       162.0        56.0                     26.0             Sci  
15       173.0        78.0                     14.0     วิทยาศาสตร์  
16       166.0        62.0                     21.0  วิศวกรรมศาสตร์  
17       161.0        50.0                     27.0     วิทยาศาสตร์  
18       174.0        76.0                      NaN      ศิลปศาสตร์  
19       168.0        64.0                     23.0     วิทยาศาสตร์

# **3.2 การทำให้ข้อมูลเป็นมาตรฐาน**

การทำให้ข้อมูลเขียนเหมือนกัน ช่วยให้เราวิเคราะห์ รวมกลุ่ม หรือทำสถิติได้ถูกต้อง

In [107]:
# ตัวอย่าง - ทำให้ชื่อคณะเป็นมาตรฐานเดียวกัน
def standardize_faculty(name):
    if name in ['วิทยาศาสตร์', 'วิทย์', 'Science', 'คณะวิทยาศาสตร์']:
        return 'วิทยาศาสตร์'
    elif name in ['ศิลปศาสตร์', 'ศิลป์', 'Arts']:
        return 'ศิลปศาสตร์'
    else:
        return name

# ใช้งาน
df['คณะ_มาตรฐาน'] = df['คณะ'].apply(standardize_faculty)

df

รหัสนักศึกษา      ชื่อ        อายุ     เพศ             คณะ   GPA  \
0          64001     สมชาย   20.000000     ชาย     วิทยาศาสตร์  3.21   
1          64002    สมหญิง   21.000000    หญิง           วิทย์  3.45   
2          64003       NaN   28.470588    Male         Science  3.67   
3          64004  สมศักดิ์   22.000000       ช  คณะวิทยาศาสตร์ -1.00   
4          64005      สมใจ   19.000000       ญ           วิศวะ  3.89   
5          64006     กานต์  150.000000     ชาย     Engineering  3.12   
6          64007     กิตติ   21.000000    หญิง        วิศวกรรม  4.50   
7          64008      จิรา   20.000000       F      ศิลปศาสตร์  2.98   
8          64009    ชนาธิป   23.000000       M            Arts  3.34   
9          64010      ดารา   21.000000    หญิง           ศิลป์  3.56   
10         64011      ธีระ   28.470588     NaN         Science  3.78   
11         64012       นภา   22.000000     ชาย           วิทย์   NaN   
12         64013     บุญมี   20.000000    หญิง           วิศวฯ  3.23   
13         64014     ปรีชา   21.000000     ชาย    Liberal Arts  3.45   
14         64015     พิมพ์   22.000000  Female             Sci  3.67   
15         64016      ภูมิ   19.000000       ช     วิทยาศาสตร์  3.89   
16         64017      มานี   20.000000       ญ  วิศวกรรมศาสตร์  3.01   
17         64018      ยุพา   21.000000    หญิง         Science  3.23   
18         64019     รัตนา   28.470588     ชาย      ศิลปศาสตร์   NaN   
19         64020     ลัดดา   22.000000     NaN         Science  3.45   

    ส่วนสูง_cm  น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์     คณะ_มาตรฐาน  
0        170.0        65.0                     15.0     วิทยาศาสตร์  
1        165.0        58.0                     20.0     วิทยาศาสตร์  
2        175.0         NaN                     25.0     วิทยาศาสตร์  
3        168.0        70.0                     10.0     วิทยาศาสตร์  
4        160.0        52.0                     30.0           วิศวะ  
5        172.0        75.0                     12.0     Engineering  
6        158.0        48.0                     35.0        วิศวกรรม  
7        180.0        82.0                     18.0      ศิลปศาสตร์  
8        163.0        55.0                     22.0      ศิลปศาสตร์  
9        167.0        60.0                     28.0      ศิลปศาสตร์  
10         NaN        68.0                      NaN     วิทยาศาสตร์  
11       171.0         NaN                     16.0     วิทยาศาสตร์  
12       169.0        72.0                     24.0           วิศวฯ  
13       164.0        54.0                     19.0    Liberal Arts  
14       162.0        56.0                     26.0             Sci  
15       173.0        78.0                     14.0     วิทยาศาสตร์  
16       166.0        62.0                     21.0  วิศวกรรมศาสตร์  
17       161.0        50.0                     27.0     วิทยาศาสตร์  
18       174.0        76.0                      NaN      ศิลปศาสตร์  
19       168.0        64.0                     23.0     วิทยาศาสตร์

# **3.3 การจัดการค่าผิดปกติ**

ข้อมูลที่ผิดจากค่าปกติอย่างชัดเจน เช่น อายุ 150 ปี อาจเกิดจากการกรอกผิด หรือข้อมูลจริงที่แตกต่างกัน

In [110]:
print("ข้อมูลต้นฉบับ")

df

ข้อมูลต้นฉบับ


รหัสนักศึกษา      ชื่อ        อายุ     เพศ             คณะ   GPA  \
0          64001     สมชาย   20.000000     ชาย     วิทยาศาสตร์  3.21   
1          64002    สมหญิง   21.000000    หญิง           วิทย์  3.45   
2          64003       NaN   28.470588    Male         Science  3.67   
3          64004  สมศักดิ์   22.000000       ช  คณะวิทยาศาสตร์ -1.00   
4          64005      สมใจ   19.000000       ญ           วิศวะ  3.89   
5          64006     กานต์  150.000000     ชาย     Engineering  3.12   
6          64007     กิตติ   21.000000    หญิง        วิศวกรรม  4.50   
7          64008      จิรา   20.000000       F      ศิลปศาสตร์  2.98   
8          64009    ชนาธิป   23.000000       M            Arts  3.34   
9          64010      ดารา   21.000000    หญิง           ศิลป์  3.56   
10         64011      ธีระ   28.470588     NaN         Science  3.78   
11         64012       นภา   22.000000     ชาย           วิทย์   NaN   
12         64013     บุญมี   20.000000    หญิง           วิศวฯ  3.23   
13         64014     ปรีชา   21.000000     ชาย    Liberal Arts  3.45   
14         64015     พิมพ์   22.000000  Female             Sci  3.67   
15         64016      ภูมิ   19.000000       ช     วิทยาศาสตร์  3.89   
16         64017      มานี   20.000000       ญ  วิศวกรรมศาสตร์  3.01   
17         64018      ยุพา   21.000000    หญิง         Science  3.23   
18         64019     รัตนา   28.470588     ชาย      ศิลปศาสตร์   NaN   
19         64020     ลัดดา   22.000000     NaN         Science  3.45   

    ส่วนสูง_cm  น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์     คณะ_มาตรฐาน  
0        170.0        65.0                     15.0     วิทยาศาสตร์  
1        165.0        58.0                     20.0     วิทยาศาสตร์  
2        175.0         NaN                     25.0     วิทยาศาสตร์  
3        168.0        70.0                     10.0     วิทยาศาสตร์  
4        160.0        52.0                     30.0           วิศวะ  
5        172.0        75.0                     12.0     Engineering  
6        158.0        48.0                     35.0        วิศวกรรม  
7        180.0        82.0                     18.0      ศิลปศาสตร์  
8        163.0        55.0                     22.0      ศิลปศาสตร์  
9        167.0        60.0                     28.0      ศิลปศาสตร์  
10         NaN        68.0                      NaN     วิทยาศาสตร์  
11       171.0         NaN                     16.0     วิทยาศาสตร์  
12       169.0        72.0                     24.0           วิศวฯ  
13       164.0        54.0                     19.0    Liberal Arts  
14       162.0        56.0                     26.0             Sci  
15       173.0        78.0                     14.0     วิทยาศาสตร์  
16       166.0        62.0                     21.0  วิศวกรรมศาสตร์  
17       161.0        50.0                     27.0     วิทยาศาสตร์  
18       174.0        76.0                      NaN      ศิลปศาสตร์  
19       168.0        64.0                     23.0     วิทยาศาสตร์

In [111]:
print(f"\nค่าเฉลี่ยอายุ (รวมค่าผิดปกติ) - {df['อายุ'].mean():.2f}")
# วิธีง่ายๆ - ใช้ IQR (Interquartile Range)
Q1 = df['อายุ'].quantile(0.25)
Q3 = df['อายุ'].quantile(0.75)
IQR = Q3 - Q1

# กำหนดขอบเขต
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# กรองค่าผิดปกติ
df_no_outliers = df[(df['อายุ'] >= lower_bound) & (df['อายุ'] <= upper_bound)]

print(f"\nขอบเขตล่าง - {lower_bound}")
print(f"ขอบเขตบน - {upper_bound}")
print(f"\nข้อมูลหลังกรองค่าผิดปกติ")
print(df_no_outliers)
print(f"\nค่าเฉลี่ยอายุ (หลังกรองค่าผิดปกติ) - {df_no_outliers['อายุ'].mean():.2f}")


ค่าเฉลี่ยอายุ (รวมค่าผิดปกติ) - 28.47

ขอบเขตล่าง - 16.625
ขอบเขตบน - 25.625

ข้อมูลหลังกรองค่าผิดปกติ
    รหัสนักศึกษา      ชื่อ  อายุ     เพศ             คณะ   GPA  ส่วนสูง_cm  \
0          64001     สมชาย  20.0     ชาย     วิทยาศาสตร์  3.21       170.0   
1          64002    สมหญิง  21.0    หญิง           วิทย์  3.45       165.0   
3          64004  สมศักดิ์  22.0       ช  คณะวิทยาศาสตร์ -1.00       168.0   
4          64005      สมใจ  19.0       ญ           วิศวะ  3.89       160.0   
6          64007     กิตติ  21.0    หญิง        วิศวกรรม  4.50       158.0   
7          64008      จิรา  20.0       F      ศิลปศาสตร์  2.98       180.0   
8          64009    ชนาธิป  23.0       M            Arts  3.34       163.0   
9          64010      ดารา  21.0    หญิง           ศิลป์  3.56       167.0   
11         64012       นภา  22.0     ชาย           วิทย์   NaN       171.0   
12         64013     บุญมี  20.0    หญิง           วิศวฯ  3.23       169.0   
13         64014     ปรีชา  21.0     ช

# **4.1 One-Hot Encoding แปลงข้อมูลประเภทเป็นตัวเลข**

AI ไม่เข้าใจข้อความ เช่น “วิทย์” หรือ “ศิลป์”
เลยต้องแปลงให้เป็นตัวเลข

In [118]:
# ทำ One-Hot Encoding
data_encoded = pd.get_dummies(df, columns=['คณะ'])

data_encoded

รหัสนักศึกษา      ชื่อ        อายุ     เพศ   GPA  ส่วนสูง_cm  น้ำหนัก_kg  \
0          64001     สมชาย   20.000000     ชาย  3.21       170.0        65.0   
1          64002    สมหญิง   21.000000    หญิง  3.45       165.0        58.0   
2          64003       NaN   28.470588    Male  3.67       175.0         NaN   
3          64004  สมศักดิ์   22.000000       ช -1.00       168.0        70.0   
4          64005      สมใจ   19.000000       ญ  3.89       160.0        52.0   
5          64006     กานต์  150.000000     ชาย  3.12       172.0        75.0   
6          64007     กิตติ   21.000000    หญิง  4.50       158.0        48.0   
7          64008      จิรา   20.000000       F  2.98       180.0        82.0   
8          64009    ชนาธิป   23.000000       M  3.34       163.0        55.0   
9          64010      ดารา   21.000000    หญิง  3.56       167.0        60.0   
10         64011      ธีระ   28.470588     NaN  3.78         NaN        68.0   
11         64012       นภา   22.000000     ชาย   NaN       171.0         NaN   
12         64013     บุญมี   20.000000    หญิง  3.23       169.0        72.0   
13         64014     ปรีชา   21.000000     ชาย  3.45       164.0        54.0   
14         64015     พิมพ์   22.000000  Female  3.67       162.0        56.0   
15         64016      ภูมิ   19.000000       ช  3.89       173.0        78.0   
16         64017      มานี   20.000000       ญ  3.01       166.0        62.0   
17         64018      ยุพา   21.000000    หญิง  3.23       161.0        50.0   
18         64019     รัตนา   28.470588     ชาย   NaN       174.0        76.0   
19         64020     ลัดดา   22.000000     NaN  3.45       168.0        64.0   

    ชั่วโมงเรียน_ต่อสัปดาห์     คณะ_มาตรฐาน  คณะ_Arts  ...  คณะ_Science  \
0                      15.0     วิทยาศาสตร์     False  ...        False   
1                      20.0     วิทยาศาสตร์     False  ...        False   
2                      25.0     วิทยาศาสตร์     False  ...         True   
3                      10.0     วิทยาศาสตร์     False  ...        False   
4                      30.0           วิศวะ     False  ...        False   
5                      12.0     Engineering     False  ...        False   
6                      35.0        วิศวกรรม     False  ...        False   
7                      18.0      ศิลปศาสตร์     False  ...        False   
8                      22.0      ศิลปศาสตร์      True  ...        False   
9                      28.0      ศิลปศาสตร์     False  ...        False   
10                      NaN     วิทยาศาสตร์     False  ...         True   
11                     16.0     วิทยาศาสตร์     False  ...        False   
12                     24.0           วิศวฯ     False  ...        False   
13                     19.0    Liberal Arts     False  ...        False   
14                     26.0             Sci     False  ...        False   
15                     14.0     วิทยาศาสตร์     False  ...        False   
16                     21.0  วิศวกรรมศาสตร์     False  ...        False   
17                     27.0     วิทยาศาสตร์     False  ...         True   
18                      NaN      ศิลปศาสตร์     False  ...        False   
19                     23.0     วิทยาศาสตร์     False  ...         True   

    คณะ_คณะวิทยาศาสตร์  คณะ_วิทยาศาสตร์  คณะ_วิทย์  คณะ_วิศวกรรม  \
0                False             True      False         False   
1                False            False       True         False   
2                False            False      False         False   
3                 True            False      False         False   
4                False            False      False         False   
5                False            False      False         False   
6                False            False      False          True   
7                False            False      False         False   
8                False            False      False         False   
9                False            False      False         False   

# **4.2 การสร้าง Feature จากวันเวลา**

In [121]:
# ตัวอย่าง - ข้อมูลการซื้อกาแฟ
coffee_data = pd.DataFrame({
    'วันเวลา': pd.to_datetime(['2024-01-15 07:30', '2024-01-15 12:00',
                               '2024-01-15 16:30', '2024-01-16 08:00']),
    'ยอดขาย': [1500, 3200, 2100, 1800]
})

# สร้าง features ใหม่
coffee_data['ชั่วโมง'] = coffee_data['วันเวลา'].dt.hour
coffee_data['วันในสัปดาห์'] = coffee_data['วันเวลา'].dt.day_name()
coffee_data['ช่วงเวลา'] = pd.cut(coffee_data['ชั่วโมง'],
                                  bins=[0, 10, 14, 18, 24],
                                  labels=['เช้า', 'กลางวัน', 'บ่าย', 'เย็น'])

print(coffee_data[['วันเวลา', 'ช่วงเวลา', 'ยอดขาย']])

              วันเวลา ช่วงเวลา  ยอดขาย
0 2024-01-15 07:30:00     เช้า    1500
1 2024-01-15 12:00:00  กลางวัน    3200
2 2024-01-15 16:30:00     บ่าย    2100
3 2024-01-16 08:00:00     เช้า    1800


# **4.3 การรวม Features**

In [123]:
# สร้าง BMI
df['BMI'] = df['น้ำหนัก_kg'] / (df['ส่วนสูง_cm']/100)**2
df['สุขภาพ'] = pd.cut(df['BMI'],
                                   bins=[0, 18.5, 25, 30, 100],
                                   labels=['ผอม', 'ปกติ', 'อ้วน', 'อ้วนมาก'])

print(df[['ชื่อ', 'BMI', 'สุขภาพ']])

        ชื่อ        BMI สุขภาพ
0      สมชาย  22.491349   ปกติ
1     สมหญิง  21.303949   ปกติ
2        NaN        NaN    NaN
3   สมศักดิ์  24.801587   ปกติ
4       สมใจ  20.312500   ปกติ
5      กานต์  25.351541   อ้วน
6      กิตติ  19.227688   ปกติ
7       จิรา  25.308642   อ้วน
8     ชนาธิป  20.700817   ปกติ
9       ดารา  21.513859   ปกติ
10      ธีระ        NaN    NaN
11       นภา        NaN    NaN
12     บุญมี  25.209201   อ้วน
13     ปรีชา  20.077335   ปกติ
14     พิมพ์  21.338211   ปกติ
15      ภูมิ  26.061679   อ้วน
16      มานี  22.499637   ปกติ
17      ยุพา  19.289379   ปกติ
18     รัตนา  25.102391   อ้วน
19     ลัดดา  22.675737   ปกติ
